src/Exception/ResponseExceptionsPool.php

```diff
@@ -0,0 +1,29 @@
+<?php declare(strict_types=1);
+
+namespace DoclerLabs\ApiClientBase\Exception;
+
+class ResponseExceptionsPool
+{
+    /** @var string[] */
+    private $responseExceptions;
+
+    public function __construct()
+    {
+        $this->responseExceptions = [
+            BadRequestResponseException::STATUS_CODE      => BadRequestResponseException::class,
+            UnauthorizedResponseException::STATUS_CODE    => UnauthorizedResponseException::class,
+            PaymentRequiredResponseException::STATUS_CODE => PaymentRequiredResponseException::class,
+            ForbiddenResponseException::STATUS_CODE       => ForbiddenResponseException::class,
+            NotFoundResponseException::STATUS_CODE        => NotFoundResponseException::class,
+        ];
+    }
+
+    public function getException(int $statusCode, string $responseBody): UnexpectedResponseException
+    {
+        if (isset($this->responseExceptions[$statusCode])) {
+            return new $this->responseExceptions[$statusCode]($responseBody);
+        }
+
+        return new UnexpectedResponseException($statusCode, $responseBody);
```

Comment:
Thanks for the refactoring!
I think that it is a factory rather then a pool. You are creating new instance of the exception class, so it should be create() method. ResponseExceptionFactory class IMO.

Reply:
I agree, done.